access/pagerduty/bot.go

```diff
@@ -39,90 +40,112 @@ func init() {
-	from        string
-	serviceID   string
+	client    *resty.Client
+	server    *WebhookServer
+	from      string
+	serviceID string
 
 	clusterName string
 }
 
-type HTTPClientImpl func(*http.Request) (*http.Response, error)
-
-func (h HTTPClientImpl) Do(req *http.Request) (*http.Response, error) {
-	return h(req)
-}
-
 func NewBot(conf PagerdutyConfig, server *WebhookServer) *Bot {
-	httpClient := &http.Client{
+	client := resty.NewWithClient(&http.Client{
 		Timeout: pdHTTPTimeout,
 		Transport: &http.Transport{
 			MaxConnsPerHost:     pdMaxConns,
 			MaxIdleConnsPerHost: pdMaxConns,
 		},
+	})
+	// APIEndpoint parameter is set only in tests
+	if conf.APIEndpoint != "" {
+		client.SetHostURL(conf.APIEndpoint)
+	} else {
+		client.SetHostURL("https://api.pagerduty.com")
 	}
+	client.SetHeader("Accept", "application/vnd.pagerduty+json;version=2")
+	client.SetHeader("Content-Type", "application/json")
+	client.SetHeader("Authorization", "Token token="+conf.APIKey)
+	client.OnBeforeRequest(func(_ *resty.Client, req *resty.Request) error {
+		req.SetError(&ErrorResult{})
+		return nil
+	})
+	client.OnAfterResponse(func(_ *resty.Client, resp *resty.Response) error {
+		if resp.IsError() {
+			result := resp.Error()
+			if result, ok := result.(*ErrorResult); ok {
+				return trace.Errorf("http error code=%v, err_code=%v, message=%v, errors=[%v]", resp.StatusCode(), result.Code, result.Message, strings.Join(result.Errors, ", "))
+			}
+			return trace.Errorf("Unknown error result %#v", result)
+		}
+		return nil
+	})
 	return &Bot{
-		httpClient:  httpClient,
-		server:      server,
-		apiEndpoint: conf.APIEndpoint,
-		apiKey:      conf.APIKey,
-		from:        conf.UserEmail,
-		serviceID:   conf.ServiceID,
+		client:    client,
+		server:    server,
+		from:      conf.UserEmail,
+		serviceID: conf.ServiceID,
 	}
 }
 
-func (b *Bot) NewClient(ctx context.Context) *pd.Client {
-	clientOpts := []pd.ClientOptions{}
-	// apiEndpoint parameter is set only in tests
-	if b.apiEndpoint != "" {
-		clientOpts = append(clientOpts, pd.WithAPIEndpoint(b.apiEndpoint))
-	}
-	client := pd.NewClient(b.apiKey, clientOpts...)
-	client.HTTPClient = HTTPClientImpl(func(r *http.Request) (*http.Response, error) {
-		return b.httpClient.Do(r.WithContext(ctx))
-	})
-	return client
-}
-
 func (b *Bot) HealthCheck(ctx context.Context) error {
-	client := b.NewClient(ctx)
-
-	if _, err := client.GetService(b.serviceID, nil); err != nil {
-		return trace.Wrap(err, "failed to fetch pagerduty service info: %v", err)
+	var result ServiceResult
+	resp, err := b.client.NewRequest().
+		SetContext(ctx).
+		SetResult(&result).
+		Get(utils.BuildURLPath("services", b.serviceID))
+	if resp != nil && resp.IsError() {
+		if contentType := resp.Header().Get("Content-Type"); contentType != "application/json" {
+			return trace.Errorf("wrong Content-Type in PagerDuty response: %q", contentType)
+		}
+		if code := resp.StatusCode(); code == http.StatusUnauthorized {
```

Comment:
What about other non-403 error codes?

Reply:
There was a function `func responseError(resp *resty.Response) error` that I used to convert an error code to Go `error`.
Now I moved it to `OnAfterResponse` callback (see `NewBot` function). So for every code > 399 the `err` will always be non-nil.

Reply:
Please leave a comment explaining this, the next person to read this code might have the same question